File: 55-jump-game/55-jump-game.cpp
class Solution {
public:
    bool jumps(int ind ,vector<int>& nums ,vector<int>& dp){
        if(ind==nums.size()-1) return true ;
        if(ind>nums.size()-1) return false ;
        
        if(dp[ind] != -1) return dp[ind] ;
        int jump = ind + nums[ind] ;
        
        for(int i=ind+1 ;i<=jump ;i++){
            bool check =jumps(i ,nums,dp) ;
            
            if(check) return dp[ind] = true ;
        }
        return dp[ind] = false ;
    }
        
    bool canJump(vector<int>& nums) {
        vector<int> dp(nums.size()+1 ,-1) ;
        
        return jumps(0 ,nums,dp) ;
    }
};

File: 55-jump-game/55-jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        int n=nums.length;
        boolean dp[]=new boolean [nums.length];
        dp[n-1]=true;
        
        
        for(int i=n-2;i>=0;i--){

            if(nums[i]+i>=n-1)
                dp[i]=true;
            else
                dp[i]=checkIfPath(i+1,i+nums[i],nums,dp);
            
        }
        
        return dp[0];
    
    }
    
    boolean checkIfPath(int start,int end,int[]nums,boolean[] dp){
        
        for(int i=start;i<=end;i++){

        if(dp[i]==true)
            return true;}
     return false;}
   
}